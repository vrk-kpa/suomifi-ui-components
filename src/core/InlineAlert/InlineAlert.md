`<InlineAlert>` is used inside the main content of a web page to convey information regarding a specific part of the site. Typically used in forms to, for example, provide information about a specific field or show centralized form errors. The content is read out to screen readers by default using `aria-live="assertive"`.

If you need to show site-wide information at the top of the page, use the <a href="#/Components/Alert">Alert</a> component instead.

Examples:

- [Basic use](./#/Components/InlineAlert?id=basic-use)
- [Warning status](./#/Components/InlineAlert?id=warning-status)
- [Error status](./#/Components/InlineAlert?id=error-status)
- [Small screen](./#/Components/InlineAlert?id=small-screen)

<div style="margin-bottom: 40px">
  [Props & methods](./#/Components/InlineAlert?id=props--methods)
</div>

### Basic use

<div style="border: 1px solid #c8cdd0; padding: 20px 20px 4px 20px; background: #eaf2fa; margin-bottom: 30px;">
#### Important!

To ensure accessibility, always wrap a dynamically appearing InlineAlert component in an aria-live region as shown in the examples below. The element which has an aria-live region must be present in the HTML document on initial page load.

</div>

```js
import { InlineAlert, Block } from 'suomifi-ui-components';

<Block aria-live="polite" style={{ width: '100%' }}>
  <InlineAlert labelText="Info">
    Make sure your name is typed exactly as it appears in your
    identification
  </InlineAlert>
</Block>;
```

### Warning status

The warning status of an InlineAlert is used for conveying information which affects users but is not absolutely fatal.

```js
import { InlineAlert, Block } from 'suomifi-ui-components';

<Block aria-live="polite" style={{ width: '100%' }}>
  <InlineAlert status="warning">
    There are grammar issues in the text
  </InlineAlert>
</Block>;
```

### Error status

The error status of an InlineAlert is used for conveying error states and issues.

```js
import { InlineAlert, Block } from 'suomifi-ui-components';

<Block aria-live="polite" style={{ width: '100%' }}>
  <InlineAlert status="error" labelText="Invalid data">
    Please fill all required fields
  </InlineAlert>
</Block>;
```

### Small screen

Set `smallScreen` to true on narrower screens (mobile devices). This applies less padding to styling.

```js
import { InlineAlert, Block } from 'suomifi-ui-components';

<Block aria-live="polite" style={{ width: '100%' }}>
  <InlineAlert labelText="Info" smallScreen>
    Make sure your name is typed exactly as it appears in your
    identification
  </InlineAlert>
</Block>;
```

### Props & methods

InlineAlert component supports [margin props](./#/Spacing/Margin%20props) for spacing.
