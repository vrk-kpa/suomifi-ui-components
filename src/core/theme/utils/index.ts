export {
  withSuomifiDefaultProps,
  themeOrTokens,
  TokensOrThemeProps,
} from './defaults';
export { focus } from './focus';
export { boxshadowOutline } from './outline';
export { margin, padding, spacingModifiers } from './spacing';
export { typographyUtils, fontSize } from './typography';
