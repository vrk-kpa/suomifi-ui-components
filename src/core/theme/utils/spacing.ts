import {
  SuomifiTheme,
  SpacingProp,
  defaultSuomifiTheme,
} from '../SuomifiTheme/SuomifiTheme';

export type SpacingWithoutInsetProp =
  | 'xxs'
  | 'xs'
  | 's'
  | 'm'
  | 'l'
  | 'xl'
  | 'xxl'
  | 'xxxl'
  | 'xxxxl'
  | '0';

const spaceVal = (theme: SuomifiTheme) => (val?: SpacingProp) => {
  if (val === '0') return '0';
  return !!val ? theme.spacing[val] : '';
};

export interface PaddingProps {
  /** Padding from theme */
  padding?: SpacingWithoutInsetProp;
  /** Padding-top from theme */
  pt?: SpacingWithoutInsetProp;
  /** Padding-right from theme */
  pr?: SpacingWithoutInsetProp;
  /** Padding-bottom from theme */
  pb?: SpacingWithoutInsetProp;
  /** Padding-left from theme */
  pl?: SpacingWithoutInsetProp;
  /** Padding on the x-axis (left & right) from theme */
  px?: SpacingWithoutInsetProp;
  /** Padding on the y-axis (top & bottom) from theme */
  py?: SpacingWithoutInsetProp;
}
export interface MarginProps {
  /** Margin from theme */
  margin?: SpacingWithoutInsetProp;
  /** Margin-top from theme */
  mt?: SpacingWithoutInsetProp;
  /** Margin-right from theme */
  mr?: SpacingWithoutInsetProp;
  /** Margin-bottom from theme */
  mb?: SpacingWithoutInsetProp;
  /** Margin-left from theme */
  ml?: SpacingWithoutInsetProp;
  /** Margin on the x-axis (left & right) from theme */
  mx?: SpacingWithoutInsetProp;
  /** Margin on the y-axis (top & bottom) from theme */
  my?: SpacingWithoutInsetProp;
}

export interface SpacingProps extends PaddingProps, MarginProps {}

export const spacingStyles = (props: SpacingProps | undefined) => {
  if (!props) return;
  const array = Object.entries(props).map(([key, value]) =>
    getSpacingStyle(defaultSuomifiTheme, key as keyof SpacingProps, value),
  );
  return Object.assign({}, ...array);
};

export const buildSpacingCSS = (props: SpacingProps | undefined): string => {
  if (!props) return '';

  const cssStyles = Object.entries(props)
    .map(([key, value]) =>
      getCSSSpacing(defaultSuomifiTheme, key as keyof SpacingProps, value),
    )
    .join('');

  return cssStyles;
};

const inlineStyle = {
  margin: 'margin',
  mt: 'marginTop',
  mr: 'marginRight',
  mb: 'marginBottom',
  ml: 'marginLeft',
  padding: 'padding',
  pt: 'paddingTop',
  pr: 'paddingRight',
  pb: 'paddingBottom',
  pl: 'paddingLeft',
};

const cssSelector = {
  margin: 'margin',
  mt: 'margin-top',
  mr: 'margin-right',
  mb: 'margin-bottom',
  ml: 'margin-left',
  padding: 'padding',
  pt: 'padding-top',
  pr: 'padding-right',
  pb: 'padding-bottom',
  pl: 'padding-left',
};

const getSpacingStyle = (
  theme: SuomifiTheme,
  key: keyof SpacingProps,
  value: SpacingProp,
) => {
  const amount = spaceVal(theme)(value);
  switch (key) {
    case 'mx':
      return { marginRight: `${amount}`, marginLeft: `${amount}` };
    case 'my':
      return { marginTop: `${amount}`, marginBottom: `${amount}` };
    case 'px':
      return { paddingRight: `${amount}`, paddingLeft: `${amount}` };
    case 'py':
      return { paddingTop: `${amount}`, paddingBottom: `${amount}` };
    case 'margin':
    case 'mt':
    case 'mr':
    case 'mb':
    case 'ml':
    case 'padding':
    case 'pt':
    case 'pr':
    case 'pb':
    case 'pl':
      return { [inlineStyle[key]]: `${amount}` };
    default:
      return '';
  }
};

const getCSSSpacing = (
  theme: SuomifiTheme,
  key: keyof SpacingProps,
  value: SpacingProp,
) => {
  const amount = spaceVal(theme)(value);
  switch (key) {
    case 'mx':
      return `margin-right: ${amount}; margin-left: ${amount};`;
    case 'my':
      return `margin-top: ${amount}; margin-bottom: ${amount};`;
    case 'px':
      return `padding-right: ${amount}; padding-left: ${amount};`;
    case 'py':
      return `padding-top: ${amount}; padding-bottom: ${amount};`;
    default:
      return `${[cssSelector[key]]}: ${amount};`;
  }
};

export const separateMarginProps = <T extends MarginProps>(
  props: T,
): [MarginProps, Omit<T, keyof MarginProps>] => {
  const { margin, my, mx, mt, mr, mb, ml, ...otherProps } = props;
  return [
    {
      ...(margin !== undefined && { margin }),
      ...(my !== undefined && { my }),
      ...(mx !== undefined && { mx }),
      ...(mt !== undefined && { mt }),
      ...(mr !== undefined && { mr }),
      ...(mb !== undefined && { mb }),
      ...(ml !== undefined && { ml }),
    },
    otherProps,
  ];
};

export const separateMarginAndPaddingProps = <T extends SpacingProps>(
  props: T,
): [SpacingProps, Omit<T, keyof SpacingProps>] => {
  const {
    margin,
    padding,
    my,
    mx,
    py,
    px,
    mt,
    mr,
    mb,
    ml,
    pt,
    pr,
    pb,
    pl,
    ...otherProps
  } = props;
  return [
    {
      ...(margin !== undefined && { margin }),
      ...(padding !== undefined && { padding }),
      ...(my !== undefined && { my }),
      ...(mx !== undefined && { mx }),
      ...(py !== undefined && { py }),
      ...(px !== undefined && { px }),
      ...(mt !== undefined && { mt }),
      ...(mr !== undefined && { mr }),
      ...(mb !== undefined && { mb }),
      ...(ml !== undefined && { ml }),
      ...(pt !== undefined && { pt }),
      ...(pr !== undefined && { pr }),
      ...(pb !== undefined && { pb }),
      ...(pl !== undefined && { pl }),
    },
    otherProps,
  ];
};
