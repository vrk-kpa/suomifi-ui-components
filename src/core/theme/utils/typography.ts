import {
  cssObjectsToCss,
  FlattenSimpleInterpolation,
  cssValueToString,
} from '../../../utils/css';
import { tokens, TypograhpyDesingTokens } from 'suomifi-design-tokens';
import { SuomifiTheme } from '../';

export type TypographyTokensAsCss = {
  [key in keyof typeof tokens.typography]: string
};

type TypographyTokensAsCssProp = {
  [key in keyof TypograhpyDesingTokens]: FlattenSimpleInterpolation
};
export interface TypographyUtil extends TypographyTokensAsCssProp {}
export class TypographyUtil {
  static instance: TypographyUtil;
  constructor(tokens: TypograhpyDesingTokens) {
    // If instance not created, does not take on account if tokens is different!
    if (!TypographyUtil.instance) {
      // Assing typographyTokens as CSS FlattenSimpleInterpolation to this object
      Object.assign(this, cssObjectsToCss(tokens));
      TypographyUtil.instance = this;
    }
    return TypographyUtil.instance;
  }
}

/**
 * Add font-declarations to typography tokens
 * @param typography Typography-tokens
 * @return typographyWithUtils Typography-tokens and typography-tokens as CSS strings
 */
export const typographyUtils = (typography: TypograhpyDesingTokens) => {
  const instance = new TypographyUtil(typography);
  Object.freeze(instance);
  return instance;
};

type ValueTypographyTokenProp = keyof typeof tokens.typography;
export const fontSize = (props: {
  theme: SuomifiTheme;
  [key: string]: any | any[];
}) => (typographyToken: ValueTypographyTokenProp) => {
  return cssValueToString(
    props.theme.values.typography[typographyToken].fontSize,
  );
};
