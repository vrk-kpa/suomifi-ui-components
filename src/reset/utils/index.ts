import { css } from 'styled-components';

const normalize = {
  html:
    'line-height: 1.15; -ms-text-size-adjust: 100%; -webkit-text-size-adjust: 100%;',
  h1: 'font-size: 2em; margin: 0.67em 0;',
  'dl dl': 'margin: 0;',
  'dl ol': 'margin: 0;',
  'dl ul': 'margin: 0;',
  'ol dl': 'margin: 0;',
  'ul dl': 'margin: 0;',
  'ol ol': 'margin: 0;',
  'ol ul': 'margin: 0;',
  'ul ol': 'margin: 0;',
  'ul ul': 'margin: 0;',
  hr: 'box-sizing: content-box; height: 0; overflow: visible;',
  main: 'display: block;',
  pre: 'font-family: monospace, monospace; font-size: 1em;',
  a: 'background-color: transparent;',
  'abbr[title]':
    'text-decoration: underline; text-decoration: underline dotted;',
  b: 'font-weight: bolder;',
  strong: 'font-weight: bolder;',
  code: 'font-family: monospace, monospace; font-size: 1em;',
  kbd: 'font-family: monospace, monospace; font-size: 1em;',
  samp: 'font-family: monospace, monospace; font-size: 1em;',
  small: 'font-size: 80%;',
  audio: 'display: inline-block;',
  video: 'display: inline-block;',
  'audio:not([controls])': 'display: none; height: 0;',
  img: 'border-style: none;',
  'svg:not(:root)': 'overflow: hidden;',
  button:
    'margin: 0;overflow: visible; text-transform: none;-webkit-appearance: button;',
  input: 'margin: 0;overflow: visible;',
  select: 'margin: 0;text-transform: none;',
  '[type="button"]': '-webkit-appearance: button;',
  '[type="reset"]': '-webkit-appearance: button;',
  '[type="submit"]': '-webkit-appearance: button;',
  fieldset: 'padding: 0.35em 0.75em 0.625em;',
  legend:
    'box-sizing: border-box; color: inherit; display: table; max-width: 100%; white-space: normal;',
  progress: 'display: inline-block; vertical-align: baseline;',
  textarea: 'margin: 0; overflow: auto;',
  '[type="checkbox"]': 'box-sizing: border-box; padding: 0;',
  '[type="radio"]': 'box-sizing: border-box; padding: 0;',
  '[type="search"]': '-webkit-appearance: textfield; outline-offset: -2px;',
  '::-webkit-inner-spin-button': 'height: auto;',
  '::-webkit-outer-spin-button': 'height: auto;',
  '::-webkit-input-placeholder': 'color: inherit; opacity: 0.54;',
  '::-webkit-search-decoration': '-webkit-appearance: none;',
  '::-webkit-file-upload-button': '-webkit-appearance: button; font: inherit;',
  '::-moz-focus-inner': 'border-style: none; padding: 0;',
  ':-moz-focusring': 'outline: 1px dotted ButtonText;',
  ':-moz-ui-invalid': 'box-shadow: none;',
  details: 'display: block;',
  dialog:
    // eslint-disable-next-line max-len
    'background-color: white; border: solid; color: black; display: block; height: -moz-fit-content; height: -webkit-fit-content; height: fit-content; left: 0; margin: auto; padding: 1em; position: absolute; right: 0; width: -moz-fit-content; width: -webkit-fit-content; width: fit-content;',
  'dialog:not([open])': 'display: none;',
  summary: 'display: list-item;',
  canvas: 'display: inline-block;',
  template: 'display: none;',
  '[hidden]': 'display: none;',
} as Record<string, string>;

const common = css`
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;

  ::before,
  ::after {
    box-sizing: border-box;
  }
`;

const resets = {
  normalize,
  common,
};

const selectorDeclaration = (selector: string, styles: string) =>
  `&${selector} {
    ${styles}
  }`;

const selectorsAndCss = (selector: string) => {
  const styles = !!normalize && normalize[selector];
  return !!styles ? selectorDeclaration(selector, styles) : '';
};

/**
 * Return selectors with resets in Emotion-css
 * @param selectors string or array of strings
 */
const resetWithSelectors = (selectors: string | string[]): string => {
  if (Array.isArray(selectors) && selectors.length > 0) {
    return `
      ${selectors.map((s) => selectorsAndCss(s)).join('')}
    `;
  }
  if (typeof selectors === 'string') {
    return selectorsAndCss(selectors);
  }
  return '';
};

export { resets, resetWithSelectors };

// TODO: Add tests
// TODO: create versioned interface for normalize.css
