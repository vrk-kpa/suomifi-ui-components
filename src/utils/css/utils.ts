import { css, FlattenSimpleInterpolation } from 'styled-components';
import { ValueUnit } from 'suomifi-design-tokens';
export { FlattenSimpleInterpolation };

export const clearfix = css`
  &:after {
    display: block;
    content: '';
    clear: both;
  }
`;

/**
 * Return FlattenSimpleInterpolation compatible string or number
 * @param cssValue compatible value or {value, unit} pair-object
 */
const cssValueToString = (cssValue: number | string | ValueAndUnit) => {
  if (!!cssValue && typeof cssValue === 'object' && 'value' in cssValue) {
    const { value, unit } = cssValue;
    return !!unit ? `${value}${unit}` : value;
  }
  return cssValue;
};

const camelToSnake = (string: string) =>
  string.replace(/[\w]([A-Z])/g, m => `${m[0]}-${m[1]}`).toLowerCase();
/**
 * Convert CSSObject to CSS FlattenSimpleInterpolation
 * @param value CSSObject
 */
export const cssObjectToCss = <T>(value: T) => css`
  ${Object.entries(value)
    .map(([key, value]) => `${camelToSnake(key)}: ${cssValueToString(value)};`)
    .join('')}
`;

/**
 * Convert CSSObject tokens to CSS FlattenSimpleInterpolation
 * @param tokens Tokens of CSSObjects
 */
export const cssObjectsToCss = <T, K extends keyof T>(tokens: T) =>
  Object.entries(tokens).reduce(
    (retObj, [key, value]) => ({
      ...retObj,
      [key]: cssObjectToCss(value),
    }),
    {} as { [key in K]: FlattenSimpleInterpolation },
  );
