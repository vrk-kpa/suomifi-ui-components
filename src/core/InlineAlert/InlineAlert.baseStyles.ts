import { css } from 'styled-components';
import { SuomifiTheme } from '../theme';
import { element, font } from '../theme/reset';
import { transparentize } from 'polished';
import { MarginProps, buildSpacingCSS } from '../theme/utils/spacing';

export const baseStyles = (
  theme: SuomifiTheme,
  globalMargins?: MarginProps,
  propMargins?: MarginProps,
) => css`
  &.fi-inline-alert {
    ${element(theme)}
    ${font(theme)('bodyTextSmall')}
    ${buildSpacingCSS(globalMargins)}
  ${buildSpacingCSS(propMargins, true)}
    width: 100%;
    padding: 5px 0px 4px 0px;
    border: 1px solid;

    & .fi-inline-alert_style-wrapper {
      display: flex;
      align-items: flex-start;
    }

    & .fi-inline-alert_icon {
      margin-left: ${theme.spacing.m};
      margin-top: ${theme.spacing.insetXl};
      min-height: 24px;
      min-width: 24px;
      height: 24px;
      width: 24px;
    }

    & .fi-inline-alert_text-content-wrapper {
      display: flex;
      flex-direction: column;
      width: 100%;
      padding: 0 ${theme.spacing.s};
      margin: ${theme.spacing.s} 0;

      & .fi-inline-alert_label {
        ${font(theme)('bodySemiBold')}
        margin-bottom: ${theme.spacing.xxs};
      }

      & .fi-inline-alert_content {
        ${font(theme)('bodyTextSmall')}
        vertical-align: middle;
      }
    }

    &.fi-inline-alert--neutral {
      background-color: ${theme.colors.infoLight1};
      border-color: ${transparentize(0.8, theme.colors.infoBase)};
      border-left: 4px solid ${theme.colors.infoBase};

      & .fi-inline-alert_text-content-wrapper {
        padding-left: ${theme.spacing.m};
      }
    }

    &.fi-inline-alert--error {
      background-color: ${theme.colors.alertLight1};
      border-color: ${transparentize(0.8, theme.colors.alertBase)};
      border-left: 4px solid ${theme.colors.alertBase};

      & .fi-inline-alert_icon--error {
        & .fi-icon-base-fill {
          fill: ${theme.colors.blackBase};
        }
      }
    }

    &.fi-inline-alert--warning {
      background-color: ${theme.colors.warningLight1};
      border-color: ${transparentize(0.8, theme.colors.warningBase)};
      border-left: 4px solid ${theme.colors.warningBase};

      & .fi-inline-alert_icon--warning {
        & .fi-icon-base-fill {
          fill: ${theme.colors.blackBase};
        }
      }
    }

    &.fi-inline-alert--small-screen {
      & .fi-inline-alert_icon {
        margin-left: ${theme.spacing.s};
      }
      & .fi-inline-alert_text-content-wrapper {
        padding: 0 ${theme.spacing.xs} 0 ${theme.spacing.xs};
      }
    }
  }
`;
