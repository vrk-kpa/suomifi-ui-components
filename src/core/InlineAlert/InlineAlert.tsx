import React, { Component, forwardRef, ReactNode } from 'react';
import { default as styled } from 'styled-components';
import classnames from 'classnames';
import { HtmlDiv, HtmlDivWithRef, HtmlDivProps } from '../../reset';
import { IconError, IconWarning } from 'suomifi-icons';
import { AutoId } from '../utils/AutoId/AutoId';
import {
  SuomifiThemeProp,
  SuomifiThemeConsumer,
  SpacingConsumer,
} from '../theme';
import {
  separateMarginProps,
  MarginProps,
  GlobalMarginProps,
} from '../theme/utils/spacing';
import { baseStyles } from './InlineAlert.baseStyles';
import { filterDuplicateKeys } from '../../utils/common/common';

const baseClassName = 'fi-inline-alert';
const inlineAlertClassNames = {
  styleWrapper: `${baseClassName}_style-wrapper`,
  content: `${baseClassName}_content`,
  label: `${baseClassName}_label`,
  textContentWrapper: `${baseClassName}_text-content-wrapper`,
  icon: `${baseClassName}_icon`,
  smallScreen: `${baseClassName}--small-screen`,
};

export interface InlineAlertProps extends HtmlDivProps, MarginProps {
  /** Style variant. Affects color and icon.
   * @default 'neutral'
   */
  status?: 'neutral' | 'warning' | 'error';
  /** Main content of the alert */
  children?: ReactNode;
  /** Toggles small screen styling */
  smallScreen?: boolean;
  /** Label for the alert */
  labelText?: ReactNode;
  /** Aria-live mode for the alert's text content and label.
   * Should not be turned completely off for dynamically appearing alerts
   * @default 'assertive'
   */
  ariaLiveMode?: 'polite' | 'assertive' | 'off';
  /** Ref is placed to the outermost div element of the component. Alternative for React `ref` attribute. */
  forwardedRef?: React.RefObject<HTMLDivElement>;
}

class BaseInlineAlert extends Component<InlineAlertProps> {
  render() {
    const {
      className,
      status = 'neutral',
      ariaLiveMode = 'assertive',
      labelText,
      children,
      smallScreen,
      id,
      forwardedRef,
      ...rest
    } = this.props;
    const [_marginProps, passProps] = separateMarginProps(rest);

    return (
      <HtmlDivWithRef
        as="section"
        {...passProps}
        className={classnames(baseClassName, className, {
          [`${baseClassName}--${status}`]: !!status,
          [inlineAlertClassNames.smallScreen]: !!smallScreen,
        })}
        style={{ ...passProps?.style }}
        ref={forwardedRef}
      >
        <HtmlDiv className={inlineAlertClassNames.styleWrapper}>
          {status === 'warning' && (
            <IconWarning
              className={classnames(inlineAlertClassNames.icon, {
                [`${inlineAlertClassNames.icon}--${status}`]: !!status,
              })}
            />
          )}
          {status === 'error' && (
            <IconError
              className={classnames(inlineAlertClassNames.icon, {
                [`${inlineAlertClassNames.icon}--${status}`]: !!status,
              })}
            />
          )}

          <HtmlDiv
            className={inlineAlertClassNames.textContentWrapper}
            id={id}
            aria-live={ariaLiveMode}
          >
            {labelText && (
              <HtmlDiv className={inlineAlertClassNames.label}>
                {labelText}
              </HtmlDiv>
            )}
            <HtmlDiv className={inlineAlertClassNames.content}>
              {children}
            </HtmlDiv>
          </HtmlDiv>
        </HtmlDiv>
      </HtmlDivWithRef>
    );
  }
}

const StyledInlineAlert = styled(
  (props: InlineAlertProps & SuomifiThemeProp & GlobalMarginProps) => {
    const { theme, globalMargins, ...passProps } = props;
    return <BaseInlineAlert {...passProps} />;
  },
)`
  ${({ theme, globalMargins, ...rest }) => {
    const [marginProps, _passProps] = separateMarginProps(rest);
    const cleanedGlobalMargins = filterDuplicateKeys(
      globalMargins.inlineAlert,
      marginProps,
    );
    return baseStyles(theme, cleanedGlobalMargins, marginProps);
  }}
`;

const InlineAlert = forwardRef<HTMLDivElement, InlineAlertProps>(
  (props: InlineAlertProps, ref: React.RefObject<HTMLDivElement>) => {
    const { id: propId, ...passProps } = props;
    return (
      <SpacingConsumer>
        {({ margins }) => (
          <SuomifiThemeConsumer>
            {({ suomifiTheme }) => (
              <AutoId id={propId}>
                {(id) => (
                  <StyledInlineAlert
                    forwardedRef={ref}
                    theme={suomifiTheme}
                    globalMargins={margins}
                    id={id}
                    {...passProps}
                  />
                )}
              </AutoId>
            )}
          </SuomifiThemeConsumer>
        )}
      </SpacingConsumer>
    );
  },
);

InlineAlert.displayName = 'InlineAlert';
export { InlineAlert };
