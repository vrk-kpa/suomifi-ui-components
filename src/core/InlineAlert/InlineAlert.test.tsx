import React from 'react';
import { render } from '@testing-library/react';
import { InlineAlert } from '../InlineAlert/InlineAlert';
import { axeTest } from '../../utils/test';

describe('children', () => {
  const alertWithElementChild = (
    <InlineAlert>
      <div>Test alert</div>
    </InlineAlert>
  );
  it('has the given content', () => {
    const { container } = render(alertWithElementChild);
    expect(container.firstChild).toHaveTextContent('Test alert');
  });

  it('should match snapshot', () => {
    const { container } = render(alertWithElementChild);
    expect(container).toMatchSnapshot();
  });
});

describe('props', () => {
  const InlineAlertComponent = (
    <InlineAlert labelText="Inline Alert label">Testcontent</InlineAlert>
  );
  it('Inline component should contain given labelText', () => {
    const { getByText } = render(InlineAlertComponent);
    const label = getByText('Inline Alert label');
    expect(label).toHaveClass('fi-inline-alert_label');
  });

  describe('Margin', () => {
    it('should have margin style from margin prop', () => {
      const { container } = render(<InlineAlert margin="xs">Test</InlineAlert>);
      expect(container.firstChild).toHaveStyle('margin: 10px');
    });

    it('should have margin prop overwritten from style prop', () => {
      const { container } = render(
        <InlineAlert margin="xs" style={{ margin: 2 }}>
          Test
        </InlineAlert>,
      );
      expect(container.firstChild).toHaveAttribute('style', 'margin: 2px;');
    });
  });
});

describe('accessibility', () => {
  const TestAlert = <InlineAlert>Testcontent</InlineAlert>;
  test('should not have basic accessibility issues', axeTest(TestAlert));
});
