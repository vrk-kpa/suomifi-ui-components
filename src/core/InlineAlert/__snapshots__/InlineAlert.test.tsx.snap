// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`children should match snapshot 1`] = `
.c2 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c2 ::before,
.c2 ::after {
  box-sizing: border-box;
}

.c0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c0 ::before,
.c0 ::after {
  box-sizing: border-box;
}

.c1.fi-inline-alert {
  color: hsl(0, 0%, 13%);
  letter-spacing: 0;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 400;
  width: 100%;
  padding: 5px 0px 4px 0px;
  border: 1px solid;
}

.c1.fi-inline-alert .fi-inline-alert_style-wrapper {
  display: flex;
  align-items: flex-start;
}

.c1.fi-inline-alert .fi-inline-alert_icon {
  margin-left: 20px;
  margin-top: 16px;
  min-height: 24px;
  min-width: 24px;
  height: 24px;
  width: 24px;
}

.c1.fi-inline-alert .fi-inline-alert_text-content-wrapper {
  display: flex;
  flex-direction: column;
  width: 100%;
  padding: 0 15px;
  margin: 15px 0;
}

.c1.fi-inline-alert .fi-inline-alert_text-content-wrapper .fi-inline-alert_label {
  letter-spacing: 0;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 600;
  margin-bottom: 5px;
}

.c1.fi-inline-alert .fi-inline-alert_text-content-wrapper .fi-inline-alert_content {
  letter-spacing: 0;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 400;
  vertical-align: middle;
}

.c1.fi-inline-alert.fi-inline-alert--neutral {
  background-color: hsl(216, 100%, 96%);
  border-color: rgba(23,112,238,0.2);
  border-left: 4px solid hsl(215, 86%, 51%);
}

.c1.fi-inline-alert.fi-inline-alert--neutral .fi-inline-alert_text-content-wrapper {
  padding-left: 20px;
}

.c1.fi-inline-alert.fi-inline-alert--error {
  background-color: hsl(0, 59%, 96%);
  border-color: rgba(174,51,45,0.2);
  border-left: 4px solid hsl(3, 59%, 43%);
}

.c1.fi-inline-alert.fi-inline-alert--error .fi-inline-alert_icon--error .fi-icon-base-fill {
  fill: hsl(0, 0%, 13%);
}

.c1.fi-inline-alert.fi-inline-alert--warning {
  background-color: hsl(42, 100%, 94%);
  border-color: rgba(246,175,9,0.2);
  border-left: 4px solid hsl(42, 93%, 50%);
}

.c1.fi-inline-alert.fi-inline-alert--warning .fi-inline-alert_icon--warning .fi-icon-base-fill {
  fill: hsl(0, 0%, 13%);
}

.c1.fi-inline-alert.fi-inline-alert--small-screen .fi-inline-alert_icon {
  margin-left: 15px;
}

.c1.fi-inline-alert.fi-inline-alert--small-screen .fi-inline-alert_text-content-wrapper {
  padding: 0 10px 0 10px;
}

<div>
  <section
    class="c0 fi-inline-alert c1 fi-inline-alert--neutral"
  >
    <div
      class="c2 fi-inline-alert_style-wrapper"
    >
      <div
        class="c2 fi-inline-alert_text-content-wrapper"
        id="3"
      >
        <div
          class="c2 fi-inline-alert_content"
        >
          <div>
            Test alert
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;
